projects/core/src/routing/store/reducers/router.reducer.ts

```diff
@@ -183,16 +190,20 @@ export class CustomSerializer
         context = { id: 'search', type: PageType.CONTENT_PAGE };
       } else if (state.data.pageLabel !== undefined) {
         context = { id: state.data.pageLabel, type: PageType.CONTENT_PAGE };
-      } else if (state.url.length > 0) {
-        context = {
-          id: '/' + state.url.map(urlSegment => urlSegment.path).join('/'),
-          type: PageType.CONTENT_PAGE
-        };
-      } else {
-        context = {
-          id: 'homepage',
-          type: PageType.CONTENT_PAGE
-        };
+      } else if (!context) {
+        if (state.url.length > 0) {
```

Comment:
Thank you for your updates. The logic looks much better now (without while.parent).
Now I can recognize 3 parts determining different 3 things:
1. is it a cms driven route (`state.firstChild.data.cxCmsRouteContext` is truthy) - I guess it is defiend only on first level and we don't need to travel down the nested routes
2. is cms required (any nested route in the line from root to leaf has `CmsPageGuard`)
3. calculate page context (doubt: using which route in the line from root to leaf? currently it's leaf, but I guess it should be root)

If i'm right, we may extract 2 methods `isCmsRequired` and `getPageContext`. While it may look like sweeping things under the rug, I think it untangles the complexity for the reader and helps us find possible bugs in our thinking. We've spent some time debugging and deducing - let's save it for the future for us and others.

If my assumptions 1-3 are correct, then I can propose such code:
```typescript
const rootState = routerState.root;
let cmsRequired;
let context;
const cxCmsRouteContext = rootState.firstChild.data.cxCmsRouteContext;
if (cxCmsRouteContext){
  cmsRequired = true;
  context = cxCmsRouteContext;
} else {
  cmsRequired = isCmsRequired(rootState); // while loop hidden inside of the method
  context = getPageContext(rootState.firstChild, params); // I assume we should use the root route, but not the leaf as we currently do
}
```